components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/AMDefaultKeyManagerImpl.java

```diff
@@ -515,86 +445,94 @@ private OAuthApplicationInfo buildDTOFromClientInfo(ClientInfo appResponse,
-        try {
-            java.net.URL keyManagerServicesURL = new java.net.URL(keyManagerServiceUrl);
-            int keyManagerPort = keyManagerServicesURL.getPort();
-            String keyManagerProtocol = keyManagerServicesURL.getProtocol();
-            String keyManagerHost = keyManagerServicesURL.getHost();
-            String keyManagerURL = keyManagerProtocol.concat("://").concat(keyManagerHost);
-            if (keyManagerPort != 0) {
-                keyManagerURL = keyManagerURL.concat(":").concat(Integer.toString(keyManagerPort));
-            }
-            String authEndpoint = tokenEndpoint.replace("/token", "");
-            revokeEndpoint = revokeEndpoint.replace("/revoke", "");
-
-            dcrClient = Feign.builder()
-                    .client(new OkHttpClient())
-                    .encoder(new GsonEncoder())
-                    .decoder(new GsonDecoder())
-                    .logger(new Slf4jLogger())
-                    .errorDecoder(new KMClientErrorDecoder())
-                    .target(DCRClient.class, keyManagerURL);
-            authClient = Feign.builder()
-                    .client(new OkHttpClient())
-                    .encoder(new GsonEncoder())
-                    .decoder(new GsonDecoder())
-                    .logger(new Slf4jLogger())
-                    .errorDecoder(new KMClientErrorDecoder()    )
-                    .encoder(new FormEncoder())
-                    .target(AuthClient.class, authEndpoint);
-            revokeClient = Feign.builder()
-                    .client(new OkHttpClient())
-                    .encoder(new GsonEncoder())
-                    .decoder(new GsonDecoder())
-                    .logger(new Slf4jLogger())
-                    .errorDecoder(new KMClientErrorDecoder()    )
-                    .encoder(new FormEncoder())
-                    .target(AuthClient.class, revokeEndpoint);
-            introspectionClient = Feign.builder()
-                    .client(new OkHttpClient())
-                    .encoder(new GsonEncoder())
-                    .decoder(new GsonDecoder())
-                    .logger(new Slf4jLogger())
-                    .errorDecoder(new KMClientErrorDecoder())
-                    .encoder(new FormEncoder())
-                    .target(IntrospectionClient.class, keyManagerURL.concat("/oauth2/introspect"));
-
-        } catch (MalformedURLException e) {
-            log.error("Error in parsing the Key Validator URL provided: " + keyManagerServiceUrl, e);
+        String dcrEndpoint;
+        if (configuration.getParameter(APIConstants.KeyManager.CLIENT_REGISTRATION_ENDPOINT) != null) {
+            dcrEndpoint = (String) configuration.getParameter(APIConstants.KeyManager.CLIENT_REGISTRATION_ENDPOINT);
+        } else {
+            dcrEndpoint = keyManagerServiceUrl.split("/" + APIConstants.SERVICES_URL_RELATIVE_PATH)[0]
+                    .concat(getTenantAwareContext().trim()).concat("/api/identity/oauth2/dcr/v1.1/register");
```

Comment:
I think having the version hardcorded in the URL could cause some maintenance issues. (will have to change this every release)

Reply:
this might be ok since we have option to read from configuration